src/Catalog/Retry.cs

```diff
@@ -0,0 +1,58 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace NuGet.Indexing
+{
+    public static class Retry
+    {
+        public static void Incremental(Action runLogic, Func<Exception, bool> shouldRetry, int maxAttempts, TimeSpan waitIncrement)
+        {
+            for (int currentAttempt = 0; currentAttempt < maxAttempts; currentAttempt++)
+            {
+                try
+                {
+                    runLogic();
+                    return;
+                }
+                catch (Exception e)
+                {
+                    if (currentAttempt < maxAttempts && shouldRetry(e))
+                    {
+                        Thread.Sleep((currentAttempt + 1) * (int)waitIncrement.TotalMilliseconds);
+                    }
+                    else
+                    {
+                        throw;
+                    }
+                }
+            }
+        }
+
+        public static async Task IncrementalAsync(Func<Task> runLogicAsync, Func<Exception, bool> shouldRetry, int maxAttempts, TimeSpan waitIncrement)
```

Comment:
Is it worth logging a custom metric for attempts > 1?  That way we could easily understand how well this solution works and also how frequently retries are necessary over time.

Reply:
Good idea. Done.